src/engraving/utests/exchangevoices_tests.cpp

```diff
@@ -104,7 +104,7 @@ TEST_F(Engraving_ExchangevoicesTests, undoChangeVoice)
     // change voice
     score->startCmd();
     score->changeSelectedNotesVoice(1);
-    score->endCmd();
+    score->endCmd(false, /*layoutAllParts = */ true);
```

Comment:
The problem with all the unit tests is that they modify the score, but they never explicitely _open_ the score (meaning, they don't call `Score::setIsOpen()`). But this is not a realistic test case: in reality, as a user, you _have_ to open a score if you want to modify it. Closed scores cannot be modified, so it is not necessary to lay them out when saving. 
So, there are two options to fix the unit tests: a) Explicitely opening the scores before modifying them, or b) Using this flag, which forces the layout of all the parts. I decided to do b) because I thought having the possibility to force a layout of all the parts might be useful for something else in future. However, I can also do a), it will take few minutes. Let me know what you think it's best to do :)

Reply:
>But this is not a realistic test case: in reality, as a user, you have to open a score if you want to modify it.

OK, how will it work in this case:  the user creates a score, opens a part, modifies it, and then closes it. After that, the user continues modifying the master score and then saves the whole project. I feel that we won't call doLayout() for the previously opened part

Reply:
So, here's what I found after lots of testing. You are totally right, we do not call doLayout() for the previously opened part before saving. As a result, yes, it _is_ possible that the saved xml will contain small mismatches between score and parts. See for example this file [part_OutOfSyncExample.mscx.zip](https://github.com/musescore/MuseScore/files/9831216/part_OutOfSyncExample.mscx.zip): the xml contains an `<accidental>` tag in the score, but not in the part.

However, I think that this is really _not_ a problem. For two reasons.
1) The mismatch can only happen with layout-dependent items, i.e. items that can be different between score and parts (for example: accidentals, beams, custom offsets etc...). The fundamental information of the score (measures, chords, notes, durations, etc) is not layout-dependent, so that will never go out of sync, so we're safe there.
2) Yes, the layout of the part is out of sync, but what can the user do with this part? a) Open it, in which case we call doLayout() and it goes back in sync with the score. b) Export it, in which case we also call doLayout(), and it goes back in sync with the score. (This can be tested on that same file: opening or exporting the Flute part, it will display a b _flat_, correctly). So, practically, the xml captures a specific moment in time where the part was out of sync, but the user will never see that, and that's what matters. In other words: we don't need the part to be in sync _in the xml_, it has to be in sync _when the user needs it_.

So, in the end, I _think_ this is all working as it should. Hope this makes sense!

Reply:
I would still consider this to be a mini bug. The current version of MuseScore might be able to deal with this slight inconsistency between the score and the parts, but I don't think there's any guarantee that future versions will deal with this appropriately. It could lead to a "corrupt score" warning dialog in the future, or incorrect behaviour if future versions allow for more differences between the score and the parts than the current version does.

Also, it's a bit of an edge case, but MSCX is an open format so MuseScore might not be the only program reading these files. We shouldn't rely on a quirk of how MuseScore reads scores to allow us to be a bit sloppy about how we write them.

Reply:
Anyway, I just thought I'd point out some potential problems here beyond what was already discussed. They appear to be edge cases though so they might be work risking if the performance gain is significant.

Reply:
That's a very valid point, which I also thought about. The way I see it is: whatever change we may do in the future, there is one immutable principle: the master score is always up to date (because it has to be open when you work on a project) and always commands the layout of a part. So I think we'll be safe. In any case, should we realize that this is a problem, reverting this change will also be painless. Let's see how it will go :)